### PR TITLE
Add Support for Microsoft Fabric / OneLake

### PR DESCRIPTION
**Pull request by vmuddassir-msft**

### CODE REVIEW

**Comment by vmuddassir-msft** [comment]
added additional onelake urls to object_store/src/azure/mod.rs for both https and abfss

this is required to fix Issue [1418](https://github.com/delta-io/delta-rs/issues/1418)

**Review by tustvold — COMMENTED** [review]
This PR appears to add support for a mix of different URLs:

* https://account.blob.fabric.microsoft.com
* https://account.dfs.fabric.microsoft.com
* abfs://account.dfs.core.windows.net
* abfs://account.dsf.fabric.microsoft.com

Could we possibly get tests for all of these, and ideally a link to some docs that explains why there are 4 different URI schemes and what the differences between them are?

**Comment by vmuddassir-msft** [comment]
Unit tests added to verify url schemes for onelake fabric url's  

https://account.blob.fabric.microsoft.com/
https://account.dfs.fabric.microsoft.com/
abfs://account.dsf.fabric.microsoft.com/

Link to [Documentation](https://learn.microsoft.com/en-us/fabric/onelake/onelake-access-api)

File: object_store/src/azure/mod.rs [hunk 1]
```diff
@@ -911,7 +928,13 @@ impl MicrosoftAzureBuilder {
             (true, url, credential, account_name)
         } else {
             let account_name = self.account_name.ok_or(Error::MissingAccount {})?;
-            let account_url = format!("https://{}.blob.core.windows.net", &account_name);
+            let account_url = if account_name.contains("onelake") {
```
Comment (tustvold) [inline object_store/src/azure/mod.rs:931]: This feels rather fragile...

**Review by tustvold — COMMENTED** [review]
I'm honestly struggling to follow how this is supposed to work, and I don't microsoft fabric, let alone a onelake endpoint setup to test this. Reading the linked doc it would suggest a fundamentally different URI scheme from what this crate expects, gone is the notion of a container and account name, instead replaced with some workspace notion?

Is there some way to interact with onelake using the normal blob storage APIs? Is this documented somewhere? Ideally we would be able to do something similar to how we support ADLS Gen 2, i.e. just treat it as normal blob storage as that is all this crate and its users are concerned with.

File: object_store/src/azure/mod.rs (hunk 1 above)
Comment (vmuddassir-msft) [reply object_store/src/azure/mod.rs:931]: @tustvold You can create an account on app.fabric.com. Once you have an account, you can follow the steps mentioned in https://learn.microsoft.com/en-us/fabric/onelake/create-lakehouse-onelake to create a lakehouse and select files to load data.

Clicking on Properties for a file used in the data load will display the Fabric URL and Azure Blob File System (ABFS) path. I have added a unit test object_store/tests/azure_object_store.rs, which uses the OneLake URL scheme for read/write operations.

image

File: object_store/src/azure/mod.rs [hunk 2]
```diff
@@ -889,7 +896,17 @@ impl MicrosoftAzureBuilder {
...
+        let use_ok_or = match &self.account_name {
+            Some(account_name) => !account_name.contains("onelake"),
+            None => true,
+        };
+
+        let container = if use_ok_or {
+            self.container_name.ok_or(Error::MissingContainerName {})?
+        } else {
+            self.container_name.unwrap_or_default()
+        };
```
Comment (tustvold) [inline object_store/src/azure/mod.rs:908]: What is the motivation for this change, AFAICT a container name is necessary to specify the workspace name?

File: object_store/tests/azure_object_store.rs [hunk 3]
```diff
@@ -0,0 +1,114 @@
+use async_trait::async_trait;
```
Comment (tustvold) [inline object_store/tests/azure_object_store.rs:1]: I think we should probably remove this before merge, but thank you for creating it to show how you are expecting to interact with this system

File: object_store/tests/azure_object_store.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+
+    let daily_store = AzureStore(
+        MicrosoftAzureBuilder::new()
+        .with_container_name("86bc63cf-5086-42e0-b16d-6bc580d1dc87")
+        .with_account("onelake")
+        .with_bearer_token_authorization("jwt-token")
+        .build()
+        .unwrap());
+    
+    let path = Path::from("17d3977c-d46e-4bae-8fed-ff467e674aed/Files/SampleCustomerList.csv");
```
Comment (tustvold) [inline object_store/tests/azure_object_store.rs:97]: This is the bit that I found rather surprising, I would have naively thought that the corollary for a container would be the files within an item, i.e. the user would just specify `SampleCustomerList.csv` in the path, with everything else provided at construction time of the builder.

This does appear to be what the docs suggest, so :shrug: To confirm if I were to perform a List of `17d3977c-d46e-4bae-8fed-ff467e674aed/Files/` it would return the full path, i.e. `17d3977c-d46e-4bae-8fed-ff467e674aed/Files/SampleCustomerList.csv` and not `SampleCustomerList.csv`?

File: object_store/src/azure/mod.rs (hunk 1 above)
Comment (tustvold) [reply object_store/src/azure/mod.rs:931]: What happens if a user has a regular storage account called onelake or containing within its name onelake, this will break it. I think we probably need some sort of `with_use_fabric_endpoint`, to explicitly gate this. This could then be set appropriately on URL extraction. What do you think?

**Review by tustvold — COMMENTED** [review]
Ok thank you for sticking with this, I'm starting to follow how this is supposed to work. The entire onelake workspace is effectively mapped as a "container", with the top-level layers of the hierarchy not being real parts of blob storage, and not responding to the APIs in a coherent manner. This does feel rather perverse though, so perhaps I am misunderstanding something... 

I've left some comments assuming the above is the correct interpretation, but feel free to correct me if I am mistaken. I am afraid I don't have the capacity, nor to be completely frank the inclination, to test yet another case of Azure fixing something that wasn't broken, so I apologise if I'm asking stupid questions.

File: object_store/src/azure/mod.rs (hunk 1 above)
Comment (vmuddassir-msft) [reply object_store/src/azure/mod.rs]: Thanks for pointing this out, I have added with_use_fabric to check if the fabric url scheme needs to be set

File: object_store/tests/azure_object_store.rs (hunk 4 above)
Comment (vmuddassir-msft) [reply object_store/tests/azure_object_store.rs:118]: Correct  , the ObjectMeta will contain the full path 
 ObjectMeta { location: Path { raw: "17d3977c-d46e-4bae-8fed-ff467e674aed/Files/SampleCustomerList.csv" }, last_modified: 2023-08-10T18:31:52Z, size: 11, e_tag: Some("0x8DB99D01129EA9D") }

image

**Comment by tustvold** [comment]
I took the liberty of pushing some minor changes to prepare this for merge

**Review by tustvold — APPROVED** [review]

**Comment by vmuddassir-msft** [comment]
@tustvold  - When will these changes to the crate be published? I noticed that the last build was released on June 6th  https://crates.io/crates/object_store/versions

**Comment by tustvold** [comment]
I intend to cut a RC today or tomorrow, with a view to publishing a release later this week

**Comment by vmuddassir-msft** [comment]
> I intend to cut a RC today or tomorrow, with a view to publishing a release later this week

Thanks - I have another PR on delta-rs that needs to be verified with the latest changes to object store crate dependency referenced in delta-rs, This is required to fix https://github.com/delta-io/delta-rs/issues/1418.

**Comment by vmuddassir-msft** [comment]
@tustvold - Any update on the release for changes to object_store crate

**Comment by tustvold** [comment]
The RC was cut on Tuesday, Apache mandates at least a 3 day voting period, and so the release should be published today
